src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -5938,6 +5983,45 @@ private void ParseTypeArgumentList(out SyntaxToken open, SeparatedSyntaxListBuil
             }
 
             close = this.EatToken(SyntaxKind.GreaterThanToken);
+
+            static bool tokenBreaksTypeArgumentList(SyntaxToken token)
+            {
+                var contextualKind = SyntaxFacts.GetContextualKeywordKind(token.ValueText);
+                switch (contextualKind)
+                {
+                    case SyntaxKind.OrKeyword:
+                    case SyntaxKind.AndKeyword:
+                    case SyntaxKind.NotKeyword:
+                        return true;
+                }
+
+                return token.Kind
+                    // Example: IEnumerable<string Method<T>()
+                    is SyntaxKind.LessThanToken
+                    // Example: Method<string(argument)
+                    or SyntaxKind.OpenParenToken
```

Comment:
i would break these into separate 'if' checks.  Different ones warrant different comments.  For example, for this, i would mention, this means we'll do a bad job with `Method<string (int a, ...` where the `(` starts a tuple type argument.  But that we're ok with taht as we don't want to look that far ahead, and we feel like it's more likely that it's a parameter list.